pyth/client/src/client_cache.rs

```diff
@@ -0,0 +1,189 @@
+                stored_vaas.insert(filename, loaded_vaas);
+            }
+        }
+
+        stored_vaas
+    }
+
+    /// Cache data.
+    pub fn store_data<I>(id: I, data: Vec<Vec<Binary>>) -> Result<(), Error>
+    where
+        I: AsRef<Path>,
+    {
+        let filename = Self::cache_filename(&id);
+
+        let cache_file = DiskPersistence::new(filename, true);
+        cache_file.save(&data)?;
+
+        Ok(())
+    }
+
+    fn cache_filename<I>(id: &I) -> PathBuf
+    where
+        I: AsRef<Path>,
+    {
+        let manifest_dir = env::var("CARGO_MANIFEST_DIR").unwrap();
+
+        let start_path = Path::new(&manifest_dir);
+
+        let workspace_root = start_path
+            .ancestors()
+            .find(|p| p.join("Cargo.lock").exists())
+            .expect("Workspace root not found");
+
+        workspace_root.join("pyth/client/testdata").join(id)
+    }
+}
+
+#[async_trait]
+impl PythClientTrait for PythClientCache {
+    type Error = error::Error;
+
+    async fn stream<I>(
+        &self,
+        ids: NonEmpty<I>,
+    ) -> Result<std::pin::Pin<Box<dyn tokio_stream::Stream<Item = Vec<Binary>> + Send>>, Self::Error>
+    where
+        I: IntoIterator + Lengthy + Send + Clone,
+        I::Item: ToString,
+    {
+        let mut stored_vaas = Self::load_or_retrieve_data(self.base_url.clone(), ids);
+        let mut index = 0;
+
+        let stream = stream! {
+            loop {
+                let vaas = stored_vaas
+                .iter_mut()
+                .filter_map(|(_, v)| v.get(index).cloned())
+                .flatten()
+                .collect::<Vec<_>>();
+                index += 1;
+
+                yield vaas;
+                sleep(Duration::from_millis(500));
+            }
+        };
+
+        Ok(Box::pin(stream))
+    }
+
+    fn get_latest_vaas<I>(&self, ids: NonEmpty<I>) -> Result<Vec<Binary>, Self::Error>
+    where
+        I: IntoIterator + Clone + Lengthy,
+        I::Item: ToString,
+    {
+        // Load the data.
+        let stored_vaas = Self::load_or_retrieve_data(self.base_url.clone(), ids.clone());
+
+        let mut return_vaas = vec![];
+
+        // For each id, try to get the vaas.
+        for id in ids.into_inner() {
+            let element = id.to_string();
+            let filename = Self::cache_filename(&element);
+
+            // Check if the vaas are stored in memory.
+            if let Some(vaas) = stored_vaas.get(&filename) {
+                if let Some(vaas) = vaas.get(self.vaas_index.load(Ordering::Acquire) as usize) {
+                    return_vaas.extend(vaas.clone());
+                }
+            } else {
+                return Err(error::Error::DataNotFound {
+                    ty: "cache",
+                    key: filename.to_string_lossy().to_string(),
+                });
+            }
+        }
+
+        self.vaas_index.fetch_add(1, Ordering::SeqCst);
+
+        Ok(return_vaas)
+    }
```

Comment:
@Zexsor33 This feels overly complex for no reason, you could just use the same pattern as the `stream` method:

```rust
    fn get_latest_vaas<I>(&self, ids: NonEmpty<I>) -> Result<Vec<Binary>, Self::Error>
    where
        I: IntoIterator + Clone + Lengthy,
        I::Item: ToString,
    {
        // Load the data.
        let stored_vaas = Self::load_or_retrieve_data(self.base_url.clone(), ids.clone());
        let index = self.vaas_index.load(Ordering::Acquire) as usize;

        let result = stored_vaas
            .iter()
            .filter_map(|(_, v)| v.get(index).cloned())
            .flatten()
            .collect::<Vec<_>>();

        self.vaas_index.fetch_add(1, Ordering::SeqCst);

        Ok(result)
    }
```

Reply:
And then you can remove the `DataNotFound` error enum from `error::Error`